winml/test/model/model_tests.cpp

```diff
@@ -0,0 +1,166 @@
+  WINML_EXPECT_NO_THROW(session = LearningModelSession(model, device));
+  WINML_EXPECT_NO_THROW(binding = LearningModelBinding(session));
+  for (int i = 0; i < m_testCase->GetDataCount(); i++) {
+
+    // Load and bind inputs
+    onnxruntime::test::HeapBuffer inputHolder;
+    std::unordered_map<std::string, Ort::Value> inputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, inputHolder, inputFeeds, true));
+    WINML_EXPECT_NO_THROW(BindInputsFromFeed(binding, inputFeeds));
+
+    // evaluate
+    LearningModelEvaluationResult results = nullptr;
+    WINML_EXPECT_NO_THROW(results = session.Evaluate(binding, L"Testing"));
+
+    // Load expected outputs
+    onnxruntime::test::HeapBuffer outputHolder;
+    std::unordered_map<std::string, Ort::Value> outputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, outputHolder, outputFeeds, false));
+
+    // compare results
+    CompareEvaluationResults(results, outputFeeds);
+  }
+}
+
+void SetTestDataPath(char* testDataPath) {
+  GetEnvironmentVariableA("WINML_TEST_DATA_PATH", testDataPath, MAX_PATH);
+  if (testDataPath[0] == 0) {
+    throw std::invalid_argument("need to set environment variable WINML_TEST_DATA_PATH");
+  }
+}
+
+// This function returns the list of all test cases inside model test collateral
+std::vector<ITestCase*> GetAllTestCases() {
+  std::vector<ITestCase*> tests;
+  std::vector<std::basic_string<PATH_CHAR_TYPE>> whitelistedTestCases;
+  double perSampleTolerance = 1e-3;
+  double relativePerSampleTolerance = 1e-3;
+  std::unordered_set<std::basic_string<ORTCHAR_T>> allDisabledTests;
+  std::vector<std::basic_string<PATH_CHAR_TYPE>> dataDirs;
+  char testDataPath[MAX_PATH];
+  SetTestDataPath(testDataPath);
+
+  for (auto& p : std::filesystem::directory_iterator(testDataPath)) {
+    if (p.is_directory()) {
+      dataDirs.push_back(p.path());
+    }
+  }
+
+  WINML_EXPECT_NO_THROW(LoadTests(dataDirs, whitelistedTestCases, perSampleTolerance, relativePerSampleTolerance,
+                                  allDisabledTests,
+                                  [&tests](std::unique_ptr<ITestCase> l) {
+                                    tests.push_back(l.get());
+                                    ownedTests.push_back(std::move(l));
+                                  }));
+  return tests;
+}
+
+// This function gets the name of the test
+static std::string GetNameOfTest(const testing::TestParamInfo<ModelTest::ParamType>& info) {
+  std::string name = "";
+  auto modelPath = std::wstring(std::get<0>(info.param)->GetModelUrl());
+  auto modelPathStr = std::string(modelPath.begin(), modelPath.end());
+  std::vector<std::string> tokenizedModelPath;
+  std::istringstream ss(modelPathStr);
+  std::string token;
+  while (std::getline(ss, token, '\\')) {
+    tokenizedModelPath.push_back(token);
+  }
+  name += tokenizedModelPath[tokenizedModelPath.size() - 2] += "_";
+  name += tokenizedModelPath[tokenizedModelPath.size() - 3] += "_";
```

Comment:
What is happening here?

Reply:
I'll add comments to describe what is happening. But basically the test collateral is structured in a way like this:

- opset7
  - squeezenet
    - model.onnx
- opset8
  - fns-candy
    - model.onnx

I want the test name to show up as : 
squeezenet_opset7_CPU
squeezenet_opset7_CPU
fns_candy_opset8_CPU
fns_candy_opset8_GPU